### PR TITLE
Use workspace versioning and relax `hashbrown`/`indexmap` deps

### LINKED ISSUE #911: Consider replacing hashbrown use in core API
### What is the expected enhancement?

Currently, some structs from `hashbrown` are used in the core API (e.g. the [`mapping` parameter of `token_swapper`](https://github.com/Qiskit/rustworkx/blob/c4f0c2e239dc45c1a5c51fc4fe0cc42be5ea9b78/rustworkx-core/src/token_swapper.rs#L410)). As a result, downstream libraries are required to use the same minor version of `hashbrown`, which has proven to be somewhat cumbersome.

Note that by switching to the standard lib equivalents, we may take _some_ perf hit, but perhaps this is worth the UX improvement.

### PR DESCRIPTION
**Pull request by jakelishman**
This commit does two things: first, switch a lot of metadata-definition to use the workspace-package inheritance allowed since Rust 1.64 (including several shared dependencies); second, relax the installation requirements on `indexmap` and `hashbrown` to allow both to use their most recent two majors.  Dependents can then install `rustworkx-core` using either (say) `hashbrown 0.13.2` or `hashbrown 0.14.0` as desired, and similar between `indexmap 1.9.3` and `indexmap 2.0.0`.

This commit leaves the lock file in the `hashbrown 0.14.0`, `indexmap 2.0.0` state, although note that some other crate dependents still use `hashbrown 0.12.3` internally, but this doesn't matter for the public API surface.  I tested it manually in the `hashbrown 0.13.2`, `indexmap 1.9.3` state by updating those two dependencies with:

    cargo update -p 'indexmap@2.0.0' --precise 1.9.3
    cargo update -p 'hashmap@0.14.0' --precise 0.13.2
    cargo update -p 'hashmap@0.12.3'

---

This commit is kind of a proof-of-concept.  I'm happy to split the different parts of it into a couple of different commits / PRs if preferred.  I only did all the workspace stuff because it was more convenient for me when playing with the versions of packages - I'm totally happy to close this PR and reopen it in a couple of different bits.

Using this, I was able to build Qiskit against `rustworkx-core =0.14.0` using both `hashbrown 0.13.2` and `hashbrown 0.14.0` with a couple of suitable `cargo update` commands.

Taken verbatim, this PR fixes #911 - it doesn't replace `hashbrown`, but it does relax the version requirements.

### CODE REVIEW

**Comment by jakelishman** [comment]
It's probably possible to avoid allowing multiple versions of `indexmap`, but when I first set out, I falsely had it in my head that `indexmap 1.9.3` exposed `hashbrown` types and so would need the flexibility, but I think I was just mistaken about that.

**Comment by jakelishman** [comment]
Last comment: I left the `hashbrown` requirement at `>=0.13`, but I didn't actually bother to try and see if it'd build at 0.12 as well.  I can if there's interest.

**Comment by mtreinish** [comment]
This is a good reminder that I forgot to forward port https://github.com/Qiskit/rustworkx/pull/929 to main. I did that on stable first for the 0.13.0 bugfix release and the msrv version constraint is lower on 0.13.x so we needed a wider hashmap range.

**Comment by jakelishman** [comment]
Hahahaha, that's funny - I guess you can just close this PR if it's already done, then.

**Comment by mtreinish** [comment]
Nah this pr is good, we need the fix on main too and this does it combined with more modern workspace usage.

File: Cargo.toml [hunk 1]
```diff
@@ -1,37 +1,60 @@
...
 
 [workspace]
 members = [
     "rustworkx-core",
 ]
 
-[lib]
-name = "rustworkx"
-crate-type = ["cdylib"]
+[workspace.package]
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply Cargo.toml:18]: I don't think you need to - `cargo publish` resolves any dynamic attributes in a `Cargo.toml` file into concrete ones.  I tried `cargo publish --dry-run -p rustworkx-core` from this branch, and the created crate has a `Cargo.toml.orig` file that looks just like `/rustworkx-core/Cargo.toml` on this branch, while the full `Cargo.toml` looks like this:
```toml
# THIS FILE IS AUTOMATICALLY GENERATED BY CARGO
#
# When uploading crates to the registry Cargo will automatically
# "normalize" Cargo.toml files for maximal compatibility
# with all versions of Cargo and also rewrite `path` dependencies
# to registry (e.g., crates.io) dependencies.
#
# If you are reading this file be aware that the original Cargo.toml
# will likely look very different (and much more reasonable).
# See Cargo.toml.orig for the original contents.

[package]
edition = "2021"
rust-version = "1.64"
name = "rustworkx-core"
version = "0.14.0"
authors = ["Matthew Treinish <mtreinish@kortar.org>"]
description = "Rust APIs used for rustworkx algorithms"
homepage = "https://github.com/Qiskit/rustworkx/tree/main/rustworkx-core"
readme = "README.md"
keywords = ["graph"]
license = "Apache-2.0"
repository = "https://github.com/Qiskit/rustworkx"

[dependencies.ahash]
version = "0.8.3"

[dependencies.fixedbitset]
version = "0.4.2"

[dependencies.hashbrown]
version = ">=0.13, <0.15"
features = ["rayon"]

[dependencies.indexmap]
version = ">=1.9, <3"
features = ["rayon"]

[dependencies.num-traits]
version = "0.2"

[dependencies.petgraph]
version = "0.6.3"

[dependencies.priority-queue]
version = "1.3"

[dependencies.rand]
version = "0.8"

[dependencies.rand_pcg]
version = "0.3"

[dependencies.rayon]
version = "1.7"

[dependencies.rayon-cond]
version = "0.3"
```

[the remark this answers is not in the document]
Comment (IvanIsCoding) [reply Cargo.toml:18]: That is fantastic, thanks for confirming it!

**Review by IvanIsCoding — APPROVED** [review]
LGTM, this will simplify keeping the dependencies between rustworkx and rustworkx-core consistent in the futre

**Pull request merged**